poolakey/src/main/java/ir/cafebazaar/poolakey/billing/connection/ServiceBillingConnection.kt

```diff
@@ -0,0 +1,292 @@
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        val intentSenderFire: (IntentSender) -> Unit = { intentSender ->
+            fragment.startIntentSenderForResult(
+                intentSender,
+                purchaseRequest.requestCode,
+                Intent(),
+                0,
+                0,
+                0,
+                null
+            )
+            PurchaseIntentCallback().apply(callback).purchaseFlowBegan.invoke()
+        }
+
+        val intentFire: (Intent) -> Unit = { intent ->
+            fragment.startActivityForResult(
+                intent,
+                purchaseRequest.requestCode
+            )
+            PurchaseIntentCallback().apply(callback).purchaseFlowBegan.invoke()
+        }
+
+        purchase(
+            purchaseRequest,
+            purchaseType,
+            callback,
+            intentSenderFire,
+            intentFire
+        )
+    }
+
+    private fun purchase(
+        purchaseRequest: PurchaseRequest,
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit,
+        fireIntentSender: (IntentSender) -> Unit,
+        fireIntent: (Intent) -> Unit
+    ) = withService {
+        purchaseFunction.function(
+            billingService = this,
+            request = PurchaseFunctionRequest(
+                purchaseRequest,
+                purchaseType,
+                callback,
+                fireIntentSender,
+                fireIntent
+            )
+        )
+    } ifServiceIsDisconnected {
+        PurchaseIntentCallback().apply(callback).failedToBeginFlow.invoke(DisconnectException())
+    }
+
+    private fun getQueryPurchasedBundle(
+        purchaseType: String,
+        continuation: String?
+    ): Bundle? {
+        return billingService?.getPurchases(
+            Billing.IN_APP_BILLING_VERSION,
+            context.packageName,
+            purchaseType,
+            continuation
+        )
+    }
+
+    private inline fun withService(
+        runOnBackground: Boolean = false,
+        crossinline service: IInAppBillingService.() -> Unit
+    ): ConnectionState {
+        return billingService?.also {
+            if (runOnBackground) {
+                backgroundThread.execute(Runnable { service.invoke(it) })
+            } else {
+                service.invoke(it)
+            }
+        }?.let { ConnectionState.Connected }
+            ?: run { ConnectionState.Disconnected }
+    }
+
+    override fun stopConnection() {
+        if (billingService != null) {
+            contextReference?.get()?.unbindService(this)
+            disconnect()
+        }
+    }
+
+    private inline infix fun ConnectionState.ifServiceIsDisconnected(block: () -> Unit) {
+        if (this is ConnectionState.Disconnected) {
+            block.invoke()
+        }
+    }
+
+    override fun onServiceDisconnected(name: ComponentName?) {
+        disconnect()
+    }
```

Comment:
The `disconnect` function inside `BillingConnection` isn't being called when service is disconnected.

Reply:
Done